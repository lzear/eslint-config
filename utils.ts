// Returns T['default'] if module has default export or T it doesn't contain default export
type ReturnDefault<T> = 'default' extends keyof T ? T['default'] : T

export let interopDefault = async <T>(
  module: Promise<T> | T,
): Promise<ReturnDefault<T>> => {
  try {
    let resolved = await module

    if (typeof resolved === 'object' && resolved !== null) {
      if ('default' in resolved && Object.keys(resolved).length === 1) {
        let defaultExport = (resolved as { default: unknown }).default as T

        if (!defaultExport) {
          // @ts-ignore
          return defaultExport
        }
        // @ts-ignore
        return defaultExport
      }
      // @ts-ignore
      return resolved
    }
    // @ts-ignore
    return resolved
  } catch (error) {
    throw new Error(`Cannot import module: ${String(error)}`)
  }
}
