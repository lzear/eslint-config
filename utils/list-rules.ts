import fs from 'node:fs'
import path from 'node:path'

import configBase from '@lzear/eslint-config'
import configReact from '@lzear/eslint-config-react'
import configTs from '@lzear/eslint-config-typescript'
import { ESLint, Linter } from 'eslint'
import _ from 'lodash'

export const listRules = async (overrideConfig: Linter.Config) => {
  const eslint = new ESLint({ overrideConfig: overrideConfig })
  const filePath = './a/somefile.js'
  const config = await eslint.calculateConfigForFile(filePath)
  const warnings: string[] = []
  const errors: string[] = []
  for (const [rule, config1] of Object.entries(config.rules)) {
    if (Array.isArray(config1)) {
      switch (config1[0]) {
        case 0: {
          break
        }
        case 1: {
          warnings.push(rule)
          break
        }
        case 2: {
          errors.push(rule)
          break
        }
        default: {
          throw new Error(`Unexpected rule configuration (${rule})`)
        }
      }
    } else {
      throw new TypeError(`Rule configuration must be an object (${rule})`)
    }
  }
  return { warnings, errors }
}

export const rulesDir = path.resolve(import.meta.dirname, '../rules')

export const allRules = async () => {
  const rulesObjs = await Promise.all(
    [configBase, configTs, configReact].map((config) =>
      listRules(config as Linter.Config),
    ),
  )
  const rules = []
  for (const { warnings, errors } of rulesObjs)
    rules.push(...warnings, ...errors)
  return _.uniq(rules).sort()
}

export const testedRules = async () => {
  const getFolders = async (
    rootPath: string,
    folders: string[] = [],
  ): Promise<string[]> => {
    const files = await fs.promises.readdir(
      path.resolve(rootPath, ...folders),
      { withFileTypes: true },
    )
    const dirs = files.filter((file) => file.isDirectory())
    if (dirs.length === 0) return [path.join(...folders)]
    const r = await Promise.all(
      dirs.map((dir) => getFolders(rootPath, [...folders, dir.name])),
    )
    return r.flat()
  }
  return getFolders(rulesDir)
}

export const undocumentedRules = async () => {
  const documented = await testedRules()
  const all = await allRules()
  return _.difference(all, documented).sort()
}
