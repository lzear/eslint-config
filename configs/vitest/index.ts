/*
  eslint perfectionist/sort-objects: [
    'error',
    {
      type: 'alphabetical',
      order: 'asc',
      partitionByNewLine: true
    }
  ]
*/

import type { Linter } from 'eslint'

import type { ConfigOptions } from '../..'

import { interopDefault } from '../../utils'

export const vitest = async (config: ConfigOptions): Promise<Linter.Config> => {
  if (!config.vitest) {
    return {}
  }

  const vitestPlugin = await interopDefault(import('@vitest/eslint-plugin'))

  const files = [
    '**/test/*.js',
    '**/test/*.cjs',
    '**/test/*.mjs',
    '**/*.test.js',
    '**/*.test.cjs',
    '**/*.test.mjs',
  ]

  if (config.typescript) {
    files.push(
      '**/test/*.ts',
      '**/test/*.cts',
      '**/test/*.mts',
      '**/*.test.ts',
      '**/*.test.cts',
      '**/*.test.mts',
    )
  }

  if (config.react) {
    files.push('**/test/*.jsx', '**/*.test.jsx')

    if (config.typescript) {
      files.push('**/test/*.tsx', '**/*.test.tsx')
    }
  }

  return {
    name: 'lzear/vitest',

    files,

    plugins: {
      vitest: vitestPlugin,
    },

    rules: { ...vitestPlugin.configs.recommended.rules },

    settings: {
      vitest: {
        typecheck: true,
      },
    },
  }
}
