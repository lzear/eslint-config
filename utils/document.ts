import fs from 'node:fs'
import path from 'node:path'

import { builtinRules } from 'eslint/use-at-your-own-risk'

import { rulesDir } from './list-rules'

const ruleMetaDic = {
  unicorn: async (rule: string) => {
    const r = await import(`eslint-plugin-unicorn/rules/${rule}`)
    return r.default.meta
  },
  js: async (rule: string) => {
    const r = builtinRules.get(rule)
    if (!r) throw new Error(`Unknown rule: ${rule}`)
    return r.meta
  },
}

const getRuleMeta = async (rule: string) => {
  const [name, domain] = rule.split('/').reverse()
  if (!name) throw new Error(`Invalid rule: ${rule}`)
  // @ts-expect-error object might miss key
  const getMeta = ruleMetaDic[domain || 'js']
  if (!getMeta) throw new Error(`Unknown domain: ${domain}`)
  return getMeta(name)
}

export const documentRule = async (rule: string) => {
  const ruleDir = path.resolve(rulesDir, rule)

  const [good, bad] = await Promise.all([
    fs.promises.readFile(path.resolve(ruleDir, 'good.ts'), 'utf8'),
    fs.promises.readFile(path.resolve(ruleDir, 'bad.ts'), 'utf8'),
  ])

  return { good, bad, meta: await getRuleMeta(rule) }
}
