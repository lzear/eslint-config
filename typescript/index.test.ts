import { Linter } from 'eslint'
import { describe, expect, it } from 'vitest'

import { listRules } from '../utils/list-rules'
import configTs from './index'

describe('typescript config', () => {
  it('matches warning rules snapshot', async () => {
    const config = await listRules(configTs as Linter.Config)
    expect(config.warnings.sort()).toMatchSnapshot()
  })

  it('matches error rules snapshot', async () => {
    const config = await listRules(configTs as Linter.Config)
    expect(config.errors.sort()).toMatchSnapshot()
  })
})
