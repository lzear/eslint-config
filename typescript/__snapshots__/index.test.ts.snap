// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`base config > matches error rules snapshot 1`] = `
[
  "constructor-super",
  "for-direction",
  "getter-return",
  "n/no-deprecated-api",
  "n/no-exports-assign",
  "n/no-extraneous-import",
  "n/no-extraneous-require",
  "n/no-missing-require",
  "n/no-process-exit",
  "n/no-unpublished-bin",
  "n/no-unpublished-import",
  "n/no-unpublished-require",
  "n/no-unsupported-features/es-builtins",
  "n/no-unsupported-features/es-syntax",
  "n/no-unsupported-features/node-builtins",
  "n/process-exit-as-throw",
  "n/shebang",
  "no-async-promise-executor",
  "no-case-declarations",
  "no-class-assign",
  "no-compare-neg-zero",
  "no-cond-assign",
  "no-const-assign",
  "no-constant-binary-expression",
  "no-constant-condition",
  "no-control-regex",
  "no-debugger",
  "no-delete-var",
  "no-dupe-args",
  "no-dupe-class-members",
  "no-dupe-else-if",
  "no-dupe-keys",
  "no-duplicate-case",
  "no-empty",
  "no-empty-character-class",
  "no-empty-pattern",
  "no-empty-static-block",
  "no-ex-assign",
  "no-extra-boolean-cast",
  "no-fallthrough",
  "no-func-assign",
  "no-global-assign",
  "no-import-assign",
  "no-invalid-regexp",
  "no-irregular-whitespace",
  "no-loss-of-precision",
  "no-misleading-character-class",
  "no-new-native-nonconstructor",
  "no-nonoctal-decimal-escape",
  "no-obj-calls",
  "no-octal",
  "no-prototype-builtins",
  "no-redeclare",
  "no-regex-spaces",
  "no-self-assign",
  "no-setter-return",
  "no-shadow-restricted-names",
  "no-sparse-arrays",
  "no-this-before-super",
  "no-undef",
  "no-unexpected-multiline",
  "no-unreachable",
  "no-unsafe-finally",
  "no-unsafe-negation",
  "no-unsafe-optional-chaining",
  "no-unused-labels",
  "no-unused-private-class-members",
  "no-unused-vars",
  "no-useless-backreference",
  "no-useless-catch",
  "no-useless-escape",
  "no-with",
  "node-import/prefer-node-protocol",
  "prefer-arrow/prefer-arrow-functions",
  "prefer-const",
  "prettier/prettier",
  "promise/always-return",
  "promise/catch-or-return",
  "promise/no-new-statics",
  "promise/no-return-wrap",
  "promise/param-names",
  "require-yield",
  "simple-import-sort/exports",
  "simple-import-sort/imports",
  "unicorn/better-regex",
  "unicorn/catch-error-name",
  "unicorn/consistent-function-scoping",
  "unicorn/empty-brace-spaces",
  "unicorn/error-message",
  "unicorn/escape-case",
  "unicorn/expiring-todo-comments",
  "unicorn/explicit-length-check",
  "unicorn/filename-case",
  "unicorn/import-style",
  "unicorn/new-for-builtins",
  "unicorn/no-anonymous-default-export",
  "unicorn/no-array-callback-reference",
  "unicorn/no-array-for-each",
  "unicorn/no-array-method-this-argument",
  "unicorn/no-array-push-push",
  "unicorn/no-array-reduce",
  "unicorn/no-await-expression-member",
  "unicorn/no-await-in-promise-methods",
  "unicorn/no-console-spaces",
  "unicorn/no-document-cookie",
  "unicorn/no-empty-file",
  "unicorn/no-for-loop",
  "unicorn/no-hex-escape",
  "unicorn/no-instanceof-array",
  "unicorn/no-invalid-remove-event-listener",
  "unicorn/no-lonely-if",
  "unicorn/no-negated-condition",
  "unicorn/no-nested-ternary",
  "unicorn/no-new-array",
  "unicorn/no-new-buffer",
  "unicorn/no-object-as-default-parameter",
  "unicorn/no-process-exit",
  "unicorn/no-single-promise-in-promise-methods",
  "unicorn/no-static-only-class",
  "unicorn/no-thenable",
  "unicorn/no-this-assignment",
  "unicorn/no-typeof-undefined",
  "unicorn/no-unnecessary-await",
  "unicorn/no-unnecessary-polyfills",
  "unicorn/no-unreadable-array-destructuring",
  "unicorn/no-unreadable-iife",
  "unicorn/no-useless-fallback-in-spread",
  "unicorn/no-useless-length-check",
  "unicorn/no-useless-promise-resolve-reject",
  "unicorn/no-useless-spread",
  "unicorn/no-useless-switch-case",
  "unicorn/no-useless-undefined",
  "unicorn/no-zero-fractions",
  "unicorn/number-literal-case",
  "unicorn/numeric-separators-style",
  "unicorn/prefer-add-event-listener",
  "unicorn/prefer-array-find",
  "unicorn/prefer-array-flat",
  "unicorn/prefer-array-flat-map",
  "unicorn/prefer-array-index-of",
  "unicorn/prefer-array-some",
  "unicorn/prefer-at",
  "unicorn/prefer-blob-reading-methods",
  "unicorn/prefer-code-point",
  "unicorn/prefer-date-now",
  "unicorn/prefer-default-parameters",
  "unicorn/prefer-dom-node-append",
  "unicorn/prefer-dom-node-dataset",
  "unicorn/prefer-dom-node-remove",
  "unicorn/prefer-dom-node-text-content",
  "unicorn/prefer-event-target",
  "unicorn/prefer-export-from",
  "unicorn/prefer-includes",
  "unicorn/prefer-keyboard-event-key",
  "unicorn/prefer-logical-operator-over-ternary",
  "unicorn/prefer-math-trunc",
  "unicorn/prefer-modern-dom-apis",
  "unicorn/prefer-modern-math-apis",
  "unicorn/prefer-module",
  "unicorn/prefer-native-coercion-functions",
  "unicorn/prefer-negative-index",
  "unicorn/prefer-node-protocol",
  "unicorn/prefer-number-properties",
  "unicorn/prefer-object-from-entries",
  "unicorn/prefer-optional-catch-binding",
  "unicorn/prefer-prototype-methods",
  "unicorn/prefer-query-selector",
  "unicorn/prefer-reflect-apply",
  "unicorn/prefer-regexp-test",
  "unicorn/prefer-set-has",
  "unicorn/prefer-set-size",
  "unicorn/prefer-spread",
  "unicorn/prefer-string-replace-all",
  "unicorn/prefer-string-slice",
  "unicorn/prefer-string-starts-ends-with",
  "unicorn/prefer-string-trim-start-end",
  "unicorn/prefer-switch",
  "unicorn/prefer-ternary",
  "unicorn/prefer-top-level-await",
  "unicorn/prefer-type-error",
  "unicorn/relative-url-style",
  "unicorn/require-array-join-separator",
  "unicorn/require-number-to-fixed-digits-argument",
  "unicorn/switch-case-braces",
  "unicorn/template-indent",
  "unicorn/text-encoding-identifier-case",
  "unicorn/throw-new-error",
  "use-isnan",
  "valid-typeof",
]
`;

exports[`base config > matches warning rules snapshot 1`] = `
[
  "promise/no-callback-in-promise",
  "promise/no-nesting",
  "promise/no-promise-in-callback",
  "promise/no-return-in-finally",
  "promise/valid-params",
]
`;
